src/client.rs

```diff
@@ -185,12 +188,117 @@ impl Stream for Events {
     }
 }
 
+pub struct Client {
+    pub framework_id: String,
+    stream_id: String,
+    pub events: Events,
+}
+
+impl fmt::Debug for Client {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Client {{ framework_id: {}, stream_id: {} }}", self.framework_id, self.stream_id)
+    }
+}
+
+header! { (MesosStreamIdHeader, "Mesos-Stream-Id") => [String] }
+
+impl Client {
+    pub fn connect(
+        handle: &Handle,
+        uri: hyper::Uri,
+        framework_info: mesos::FrameworkInfo,
+    ) -> Box<Future<Item = Self, Error = failure::Error>> {
+        // Mesos subscribe call
+        let call = Self::subscribe(framework_info);
+        let request = Self::request_for(uri, call);
+
+        // Call Mesos
+        let http_client = hyper::Client::new(&handle);
+        let s = http_client
+            .request(request)
+            .map_err(failure::Error::from)
+            .and_then(move |res: hyper::Response| {
+                println!("Response status: {}", res.status());
+
+                let stream_id = if let Some(header) = res.headers().get::<MesosStreamIdHeader>() {
+                    future::ok(header.0.clone())
+                } else {
+                    // TODO: Use different error type.
+                    future::err(format_err!("Missing Mesos-Stream-Id header."))
+                };
+                let events = Events::new(res.body()).into_future().map_err(|(err, _)| {
+                    failure::Error::from(err)
+                });
+                events.join(stream_id)
+            })
+            .and_then(|((subscribed_event, events), stream_id)| {
+                future::result(Self::client_from(subscribed_event, events, stream_id))
+            });
+        Box::new(s)
+    }
+
+    pub fn subscribe(framework_info: mesos::FrameworkInfo) -> scheduler::Call {
+        let mut call = scheduler::Call::new();
+        let mut subscribe = scheduler::Call_Subscribe::new();
+        subscribe.set_framework_info(framework_info);
+        call.set_subscribe(subscribe);
+        call.set_field_type(scheduler::Call_Type::SUBSCRIBE);
+        call
+    }
+
+    /// Builds a Hyper Request object for given URI and Mesos scheduler call.
+    pub fn request_for(uri: hyper::Uri, call: scheduler::Call) -> hyper::Request {
+        let mut request = hyper::Request::new(hyper::Method::Post, uri);
+        // TODO: construct media type once
+        let protobuf_media_type = "application/x-protobuf".parse::<mime::Mime>().unwrap();
```

Comment:
lazy_static is your friend

Reply:
Done. At some point it should just be a `const mime::Mime` such as `mime::Text_PLAIN`